core/src/main/kotlin/org/eln2/data/MultiMap.kt

```diff
@@ -0,0 +1,197 @@
+package org.eln2.data
+
+/**
+ * A helper class for representing a mapping as a Map.Entry.
+ */
+data class MultiMapEntry<K, V>(override val key: K, override val value: V): Map.Entry<K, V>
+
+/**
+ * A MultiMap: a data structure that maps one key (K) to zero or more values (V).
+ *
+ * This is in contrast to the usual Map, which maps one key to at most one value.
+ *
+ */
+interface MultiMap<K, V> {
+	/**
+	 * Tests whether a key [k] is in the MultiMap.
+	 *
+	 * This is true if k is associated with at least one value.
+	 */
+	operator fun contains(k: K): Boolean = get(k).isNotEmpty()
+
+	/**
+	 * Get the set of all values associated with key [k].
+	 */
+	operator fun get(k: K): Set<V>
+
+	/**
+	 * An iterator over every Entry<K, Set<V>>, of size keySize.
+	 */
+	val keyMapping: Iterable<Map.Entry<K, Set<V>>>
+	/**
+	 * The number of keys, value sets, and keyMapping entries in this MultiMap.
+	 */
+	val keyMappingSize: Int get() = keyMapping.map { 1 }.sum()
+	/**
+	 * The set of all keys.
+	 */
+	val keys: Set<K> get() = keyMapping.map {(k, _) -> k }.toSet()
```

Comment:
Doesn't this already have .keys()?

Also, newline.

Reply:
I don't think you addressed this.

Reply:
You're right, I missed this :)

`keyMapping: Iterable<Map.Entry<K, Set<V>>>` which... doesn't have `keys` on it, if that's what you're asking; you _could construct a `Map` out of it proper and take its keys, but I was avoiding round-tripping through another object for that, especially since the `Map`-dependent code is in the concrete implementation.

At least, I think that's what you're asking? Let me know if you meant something else :)

Reply:
Well, keyMapping is constructing an iterable from a concrete map which does have .keys(). But let's leave that as a todo for the future.